### ISSUE TITLE
different results between the CPUExecutionProvider and CUDAExecutionProvider when onnxruntime executes a valid model

### ISSUE THREAD

**Issue by coffezhou** [microsoft/onnxruntime#25051]
### Describe the issue

For the following onnx model,

![Image](https://github.com/user-attachments/assets/86b477d0-43f7-4157-a325-81581f82776d)
the results of onnxruntime with the CPUExecutionProvider are as follows:
```c
ONNXRuntime:
 [array([[[[0.15210256, 0.15210256, 0.15210256, ..., 0.15210256,
          0.15210256, 0.15210256],
         [0.15210256, 0.15210256, 0.15210256, ..., 0.15210256,
          0.15210256, 0.15210256],
         [0.15210256, 0.15210256, 0.15210256, ..., 0.15210256,
          0.15210256, 0.15210256],
         ...,
         [0.15210256, 0.15210256, 0.15210256, ..., 0.15210256,
          0.15210256, 0.15210256],
         [0.15210256, 0.15210256, 0.15210256, ..., 0.15210256,
          0.15210256, 0.15210256],
         [0.15210256, 0.15210256, 0.15210256, ..., 0.15210256,
          0.15210256, 0.15210256]],

        [[0.15210256, 0.15210256, 0.15210256, ..., 0.15210256,
          0.15210256, 0.15210256],
         [0.15210256, 0.15210256, 0.15210256, ..., 0.15210256,
          0.15210256, 0.15210256],
         [0.15210256, 0.15210256, 0.15210256, ..., 0.15210256,
          0.15210256, 0.15210256],
         ...,
         [0.15210256, 0.15210256, 0.15210256, ..., 0.15210256,
          0.15210256, 0.15210256],
         [0.15210256, 0.15210256, 0.15210256, ..., 0.15210256,
          0.15210256, 0.15210256],
         [0.15210256, 0.15210256, 0.15210256, ..., 0.15210256,
          0.15210256, 0.15210256]],

        [[0.69579494, 0.69579494, 0.69579494, ..., 0.69579494,
          0.69579494, 0.69579494],
         [0.69579494, 0.69579494, 0.69579494, ..., 0.69579494,
          0.69579494, 0.69579494],
         [0.69579494, 0.69579494, 0.69579494, ..., 0.69579494,
          0.69579494, 0.69579494],
         ...,
         [0.69579494, 0.69579494, 0.69579494, ..., 0.69579494,
          0.69579494, 0.69579494],
         [0.69579494, 0.69579494, 0.69579494, ..., 0.69579494,
          0.69579494, 0.69579494],
         [0.69579494, 0.69579494, 0.69579494, ..., 0.69579494,
          0.69579494, 0.69579494]]]],
      shape=(1, 3, 224, 224), dtype=float32)]
```
However, I run onnxruntime with the CUDAExecutionProvider, the results are all nan:
```c
ONNXRuntime:
 [array([[[[nan, nan, nan, ..., nan, nan, nan],
         [nan, nan, nan, ..., nan, nan, nan],
         [nan, nan, nan, ..., nan, nan, nan],
         ...,
         [nan, nan, nan, ..., nan, nan, nan],
         [nan, nan, nan, ..., nan, nan, nan],
         [nan, nan, nan, ..., nan, nan, nan]],

        [[nan, nan, nan, ..., nan, nan, nan],
         [nan, nan, nan, ..., nan, nan, nan],
         [nan, nan, nan, ..., nan, nan, nan],
         ...,
         [nan, nan, nan, ..., nan, nan, nan],
         [nan, nan, nan, ..., nan, nan, nan],
         [nan, nan, nan, ..., nan, nan, nan]],

        [[nan, nan, nan, ..., nan, nan, nan],
         [nan, nan, nan, ..., nan, nan, nan],
         [nan, nan, nan, ..., nan, nan, nan],
         ...,
         [nan, nan, nan, ..., nan, nan, nan],
         [nan, nan, nan, ..., nan, nan, nan],
         [nan, nan, nan, ..., nan, nan, nan]]]],
      shape=(1, 3, 224, 224), dtype=float32)]
```
I am not sure that this issue is due to the difference between the CPU and CUDA.

### Environment
OS: Ubuntu 20.04
onnxruntime: 1.23.0.dev20250515001
CUDA: cuda-12.2.2::cuda-toolkit
CUDNN: 9.1.1.17
NVIDIA GPU: GeForce RTX 3080
NVIDIA Driver Version: 535.183.01
Python Version: 3.12.9

### Steps to reproduce
This bug can be reproduced by the following code with the model in the attachment.
```python
from typing import Dict, List, Literal, Optional
import sys
import os

import numpy as np
import onnx
import onnxruntime
from onnx import ModelProto, TensorProto, helper, mapping

import pickle

def test():
    
    onnx_model = onnx.load("a1475.onnx")
    print(onnx_model.opset_import[0].version)

    with open("inputs.pkl", "rb") as fp:
        inputs = pickle.load(fp)

    ort_session = onnxruntime.InferenceSession(
            onnx_model.SerializeToString(), providers=["CPUExecutionProvider"]
        )
    ort_output = ort_session.run([], inputs)
    
    print("ONNXRuntime:\n", ort_output)
    
    #--------------------------------------------
    
    ort_session = onnxruntime.InferenceSession(
            onnx_model.SerializeToString(), providers=["CUDAExecutionProvider"]
        )
    ort_output = ort_session.run([], inputs)
    
    print("ONNXRuntime:\n", ort_output)
    
if __name__ == "__main__":
    test()

```

[testcase.zip](https://github.com/user-attachments/files/20727054/testcase.zip)

### Platform

Linux

### OS Version

Ubuntu 20.04

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

1.23.0.dev20250515001

### ONNX Runtime API

Python

### Architecture

X64

### Execution Provider

Default CPU, CUDA

### Execution Provider Library Version

CUDA: cuda-12.2.2::cuda-toolkit CUDNN: 9.1.1.17

**Comment by tianleiwu** [comment]
Could you try build from source with `--cmake_extra_defines onnxruntime_DEBUG_NODE_INPUTS_OUTPUTS=1`. Then you can set `ORT_DEBUG_NODE_IO_DUMP_INPUT_DATA=1` and `ORT_DEBUG_NODE_IO_DUMP_OUTPUT_DATA=1` before running test. It's able to see which node in CUDA provider generates NaN.

See https://onnxruntime.ai/docs/build/eps.html#cuda

**Comment by coffezhou** [comment]
> Could you try build from source with `--cmake_extra_defines onnxruntime_DEBUG_NODE_INPUTS_OUTPUTS=1`. Then you can set `ORT_DEBUG_NODE_IO_DUMP_INPUT_DATA=1` and `ORT_DEBUG_NODE_IO_DUMP_OUTPUT_DATA=1` before running test. It's able to see which node in CUDA provider generates NaN.
> 
> See https://onnxruntime.ai/docs/build/eps.html#cuda

Hi @tianleiwu ,
Thanks for your reply! I have verified that the different results are introduced by the Erf operator. When I delete Erf and Softmax, the results are identical.
```c
[array([[[[ 0.5       ,  0.5       ,  0.5       , ...,  0.5       ,
           0.5       ,  0.5       ],
         [ 0.5       ,  0.5       ,  0.5       , ...,  0.5       ,
           0.5       ,  0.5       ],
         [ 0.5       ,  0.5       ,  0.5       , ...,  0.5       ,
           0.5       ,  0.5       ],
         ...,
         [ 0.5       ,  0.5       ,  0.5       , ...,  0.5       ,
           0.5       ,  0.5       ],
         [ 0.5       ,  0.5       ,  0.5       , ...,  0.5       ,
           0.5       ,  0.5       ],
         [ 0.5       ,  0.5       ,  0.5       , ...,  0.5       ,
           0.5       ,  0.5       ]],

        [[-1.274278  , -4.081801  , -2.8481493 , ..., -3.6959515 ,
          -2.5275178 , -4.204624  ],
         [-0.98449033, -3.1676192 , -0.8963538 , ..., -2.759856  ,
          -2.6241748 , -2.8154266 ],
         [-1.3715359 , -0.81958896, -2.7713897 , ..., -0.90578246,
          -2.5001547 , -1.016683  ],
         ...,
         [-3.961127  , -1.5257069 , -1.8830471 , ..., -1.0672092 ,
          -2.5570152 , -2.0075402 ],
         [-1.109859  , -1.8480891 , -1.423282  , ..., -1.509791  ,
          -3.7576206 , -2.9536216 ],
         [-2.4129372 , -3.186962  , -0.94226515, ..., -1.698816  ,
          -6.174673  , -1.6355606 ]],

        [[        nan,         nan,         nan, ...,         nan,
                  nan,         nan],
         [        nan,         nan,         nan, ...,         nan,
                  nan,         nan],
         [        nan,         nan,         nan, ...,         nan,
                  nan,         nan],
         ...,
         [        nan,         nan,         nan, ...,         nan,
                  nan,         nan],
         [        nan,         nan,         nan, ...,         nan,
                  nan,         nan],
         [        nan,         nan,         nan, ...,         nan,
                  nan,         nan]]]],
      shape=(1, 3, 224, 224), dtype=float32)]
```
However, if I add the Erf operator, the results of the CUDA backend are as follows:
```c
[array([[[[      nan,       nan,       nan, ...,       nan,       nan,
                nan],
         [      nan,       nan,       nan, ...,       nan,       nan,
                nan],
         [      nan,       nan,       nan, ...,       nan,       nan,
                nan],
         ...,
         [      nan,       nan,       nan, ...,       nan,       nan,
                nan],
         [      nan,       nan,       nan, ...,       nan,       nan,
                nan],
         [      nan,       nan,       nan, ...,       nan,       nan,
                nan]],

        [[0.5204999, 0.5204999, 0.5204999, ..., 0.5204999, 0.5204999,
          0.5204999],
         [0.5204999, 0.5204999, 0.5204999, ..., 0.5204999, 0.5204999,
          0.5204999],
         [0.5204999, 0.5204999, 0.5204999, ..., 0.5204999, 0.5204999,
          0.5204999],
         ...,
         [0.5204999, 0.5204999, 0.5204999, ..., 0.5204999, 0.5204999,
          0.5204999],
         [0.5204999, 0.5204999, 0.5204999, ..., 0.5204999, 0.5204999,
          0.5204999],
         [0.5204999, 0.5204999, 0.5204999, ..., 0.5204999, 0.5204999,
          0.5204999]],

        [[      nan,       nan,       nan, ...,       nan,       nan,
                nan],
         [      nan,       nan,       nan, ...,       nan,       nan,
                nan],
         [      nan,       nan,       nan, ...,       nan,       nan,
                nan],
         ...,
         [      nan,       nan,       nan, ...,       nan,       nan,
                nan],
         [      nan,       nan,       nan, ...,       nan,       nan,
                nan],
         [      nan,       nan,       nan, ...,       nan,       nan,
                nan]]]], shape=(1, 3, 224, 224), dtype=float32)]
```
this is different from those of the CPU backend:
```c
[array([[[[-1.       , -1.       , -1.       , ..., -1.       ,
          -1.       , -1.       ],
         [-1.       , -1.       , -1.       , ..., -1.       ,
          -1.       , -1.       ],
         [-1.       , -1.       , -1.       , ..., -1.       ,
          -1.       , -1.       ],
         ...,
         [-1.       , -1.       , -1.       , ..., -1.       ,
          -1.       , -1.       ],
         [-1.       , -1.       , -1.       , ..., -1.       ,
          -1.       , -1.       ],
         [-1.       , -1.       , -1.       , ..., -1.       ,
          -1.       , -1.       ]],

        [[ 0.5204999,  0.5204999,  0.5204999, ...,  0.5204999,
           0.5204999,  0.5204999],
         [ 0.5204999,  0.5204999,  0.5204999, ...,  0.5204999,
           0.5204999,  0.5204999],
         [ 0.5204999,  0.5204999,  0.5204999, ...,  0.5204999,
           0.5204999,  0.5204999],
         ...,
         [ 0.5204999,  0.5204999,  0.5204999, ...,  0.5204999,
           0.5204999,  0.5204999],
         [ 0.5204999,  0.5204999,  0.5204999, ...,  0.5204999,
           0.5204999,  0.5204999],
         [ 0.5204999,  0.5204999,  0.5204999, ...,  0.5204999,
           0.5204999,  0.5204999]],

        [[-1.       , -1.       , -1.       , ..., -1.       ,
          -1.       , -1.       ],
         [-1.       , -1.       , -1.       , ..., -1.       ,
          -1.       , -1.       ],
         [-1.       , -1.       , -1.       , ..., -1.       ,
          -1.       , -1.       ],
         ...,
         [-1.       , -1.       , -1.       , ..., -1.       ,
          -1.       , -1.       ],
         [-1.       , -1.       , -1.       , ..., -1.       ,
          -1.       , -1.       ],
         [-1.       , -1.       , -1.       , ..., -1.       ,
          -1.       , -1.       ]]]],
      shape=(1, 3, 224, 224), dtype=float32)]
```
Maybe there are some different to process the NaN values between the CPUExecutionProvider and CUDAExecutionProvider. I also further analysis which operator produces the the NaN values, and find the results of the BatchNormalization operator also contain NaN values. This is vary strange.

The following attachment is the reduced onnx models.
[models.zip](https://github.com/user-attachments/files/20941914/models.zip)

**Comment by tianleiwu** [comment]
It is not likely caused by Erf since cuda use [erff](https://docs.nvidia.com/cuda/cuda-math-api/cuda_math_api/group__CUDA__MATH__SINGLE.html#_CPPv44erfff), which will return NaN when input is NaN. It is elementwise operator, so a whole line of NaN means the input to Erf also has a whole line of NaN.

You can set another environment variable like [ORT_DEBUG_NODE_IO_DUMP_STATISTICS_DATA](https://github.com/microsoft/onnxruntime/blob/849eee8370ac9c862682deba32bc5721ca1f994c/onnxruntime/core/framework/debug_node_inputs_outputs_utils.h#L43C46-L43C84)=1. It will show statistics. Then it is easy to find the first node generates NaN.
